File: youzi_v2/routers/shipment_sla.py
from fastapi import APIRouter, HTTPException, Query

from ..context import *

router = APIRouter()


@router.get("/api/v1/channels/{code}/sla-rules")
def list_channel_sla_rules(code: str):
    if channels_repo.get_row(code) is None:
        raise HTTPException(status_code=404, detail="渠道不存在")
    return {"items": channel_sla_rules_repo.list_by_channel(code)}


@router.put("/api/v1/channels/{code}/sla-rules")
def upsert_channel_sla_rule(code: str, body: ChannelSlaRuleIn):
    if channels_repo.get_row(code) is None:
        raise HTTPException(status_code=404, detail="渠道不存在")
    try:
        row = channel_sla_rules_repo.upsert_rule(
            channel_code=code,
            estimated_days=body.estimated_days,
            carrier_code=body.carrier_code,
            start_field=body.start_field,
            warning_days=body.warning_days,
            severe_overdue_days=body.severe_overdue_days,
            enabled=body.enabled,
            note=body.note,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    return row


@router.delete("/api/v1/channels/{code}/sla-rules/{rule_id}", status_code=204)
def delete_channel_sla_rule(code: str, rule_id: str):
    rules = channel_sla_rules_repo.list_by_channel(code)
    if not any(r["id"] == rule_id for r in rules):
        raise HTTPException(status_code=404, detail="规则不存在")
    channel_sla_rules_repo.delete_rule(rule_id)
    return Response(status_code=204)


@router.get("/api/v1/shipment-sla-alerts")
def list_shipment_sla_alerts(
    scope: str = Query("todo", description="todo=默认待办（未签收且 open/acknowledged）；all=含历史"),
    risk_level: str | None = Query(None, alias="riskLevel"),
    alert_type: str | None = Query(None, alias="alertType"),
    status: str | None = None,
    has_exception: bool | None = Query(None, alias="hasException"),
    exception_code: str | None = Query(None, alias="exceptionCode"),
    channel_code: str | None = Query(None, alias="channelCode"),
    carrier_code: str | None = Query(None, alias="carrierCode"),
    customer: str | None = None,
    search: str | None = None,
    limit: int = 50,
    offset: int = 0,
):
    return shipment_sla_alerts_repo.list_alerts(
        scope=scope,
        risk_level=risk_level,
        alert_type=alert_type,
        status=status,
        has_exception=has_exception,
        exception_code=exception_code,
        channel_code=channel_code,
        carrier_code=carrier_code,
        customer=customer,
        search=search,
        limit=limit,
        offset=offset,
    )


@router.get("/api/v1/shipment-sla-alerts/summary")
def get_shipment_sla_alerts_summary():
    return shipment_sla_alerts_repo.summary_counts()


@router.get("/api/v1/shipment-sla-alerts/todo-notifications")
def list_shipment_sla_todo_notifications(limit: int = 20):
    lim = max(1, min(limit, 50))
    return {
        "items": shipment_sla_alerts_repo.list_todo_notifications(limit=lim),
        "pendingCount": shipment_sla_alerts_repo.count_todo(),
    }


@router.post("/api/v1/shipment-sla-alerts/evaluate", response_model=ShipmentSlaScanResult)
def evaluate_shipment_sla_alerts():
    result = scan_shipment_sla_alerts(
        shipment_sla_alerts_repo,
        channel_sla_rules_repo,
        shipment_exception_followup_repo,
        force=True,
        trigger="manual",
    )
    return ShipmentSlaScanResult(**result)


@router.post("/api/v1/shipment-sla-alerts/{alert_id}/follow-up")
def follow_up_shipment_sla_alert(alert_id: str):
    if not shipment_sla_alerts_repo.follow_up(alert_id):
        raise HTTPException(status_code=404, detail="预警不存在或不可跟进")
    return {"ok": True}


@router.post("/api/v1/shipment-sla-alerts/{alert_id}/acknowledge")
def acknowledge_shipment_sla_alert(alert_id: str):
    if not shipment_sla_alerts_repo.follow_up(alert_id):
        raise HTTPException(status_code=404, detail="预警不存在或不可跟进")
    return {"ok": True}


@router.post("/api/v1/shipment-sla-alerts/{alert_id}/resolve")
def resolve_shipment_sla_alert(alert_id: str):
    if not shipment_sla_alerts_repo.resolve_alert(alert_id):
        raise HTTPException(status_code=404, detail="预警不存在或不可解决")
    return {"ok": True}


@router.post("/api/v1/shipment-sla-alerts/{alert_id}/ignore")
def ignore_shipment_sla_alert(alert_id: str):
    if not shipment_sla_alerts_repo.ignore(alert_id):
        raise HTTPException(status_code=404, detail="预警不存在或已结案")
    return {"ok": True}


@router.post("/api/v1/shipment-sla-alerts/{alert_id}/convert")
def convert_shipment_sla_alert(alert_id: str, body: ShipmentSlaAlertConvertIn):
    alert = shipment_sla_alerts_repo.get_alert(alert_id)
    if not alert:
        raise HTTPException(status_code=404, detail="预警不存在")
    if alert["status"] not in ("open", "acknowledged"):
        raise HTTPException(status_code=400, detail="该预警已不可转异常")

    shipment_no = alert["shipmentNo"]
    open_result = shipment_exceptions_repo.open_for_shipment_nos(
        [shipment_no],
        body.exception_code,
        note=body.note or None,
    )
    if open_result.get("errors"):
        err = open_result["errors"][0]
        raise HTTPException(status_code=400, detail=err.get("message", "转异常失败"))
    if not open_result.get("opened") and open_result.get("skipped"):
        pass

    events = shipment_exceptions_repo.list_by_shipment_no(shipment_no, limit=1)
    event_id = events["items"][0]["id"] if events.get("items") else ""

    if not shipment_sla_alerts_repo.mark_converted(
        alert_id,
        exception_code=body.exception_code,
        event_id=event_id,
    ):
        raise HTTPException(status_code=400, detail="更新预警状态失败")
    return {"ok": True, "eventId": event_id}


@router.patch("/api/v1/shipment-sla-alerts/{alert_id}/note")
def update_shipment_sla_alert_note(alert_id: str, body: ShipmentSlaAlertNoteIn):
    if not shipment_sla_alerts_repo.update_note(alert_id, body.note):
        raise HTTPException(status_code=404, detail="预警不存在")
    return {"ok": True}

File: youzi_v2/frontend/src/api/shipmentSla.ts
import { api } from './client'

export type SlaRiskLevel = 'warning_soon' | 'overdue' | 'severe_overdue'
export type SlaAlertStatus = 'open' | 'acknowledged' | 'converted' | 'resolved' | 'ignored'

export interface ChannelSlaRule {
  id: string
  channelCode: string
  carrierCode: string
  startField: string
  estimatedDays: number
  warningDays: number
  severeOverdueDays: number
  enabled: boolean
  note: string
  createdTime: string
  updatedTime: string
}

export interface ChannelSlaRulePayload {
  estimatedDays: number
  carrierCode?: string
  startField?: string
  warningDays?: number
  severeOverdueDays?: number
  enabled?: boolean
  note?: string
}

export interface ShipmentSlaAlert {
  id: string
  shipmentId: string
  shipmentNo: string
  alertType: string
  riskLevel: SlaRiskLevel
  status: SlaAlertStatus
  severity: string
  ruleId: string
  ruleScope: string
  channelCode: string
  carrierCode: string
  startField: string
  startTime: string
  dueDate: string
  warningDate: string
  convertedExceptionCode: string
  convertedEventId: string
  acknowledgedTime: string | null
  resolvedTime: string | null
  ignoredTime: string | null
  followUpCount: number
  lastFollowUpTime: string | null
  lastFollowUpDaysAgo: number | null
  owner: string
  note: string
  eventKey: string
  createdTime: string
  updatedTime: string
  daysUntilDue: number | null
  overdueDays: number
  daysInTransit: number | null
  estimatedDays: number | null
  customer?: string | null
  destinationPort?: string | null
  atd?: string | null
  warehouseEntryTime?: string | null
  expectedDeliveryTime?: string | null
  deliveredTime?: string | null
  exceptionCode?: string | null
  exceptionNameZh?: string | null
  exceptionDurationSeconds?: number | null
  exceptionDurationLabel?: string | null
  latestTrackingDesc?: string | null
  latestTrackingTime?: string | null
  channelNameZh?: string | null
  carrierNameZh?: string | null
}

export interface ShipmentSlaAlertListResponse {
  items: ShipmentSlaAlert[]
  total: number
  limit: number
  offset: number
}

export interface ShipmentSlaSummary {
  pendingOpen: number
  severeOverdue: number
  overdue: number
  warningSoon: number
  currentExceptions: number
}

export async function listChannelSlaRules(channelCode: string): Promise<{ items: ChannelSlaRule[] }> {
  return api(`/api/v1/channels/${encodeURIComponent(channelCode)}/sla-rules`)
}

export async function upsertChannelSlaRule(
  channelCode: string,
  payload: ChannelSlaRulePayload,
): Promise<ChannelSlaRule> {
  return api(`/api/v1/channels/${encodeURIComponent(channelCode)}/sla-rules`, {
    method: 'PUT',
    body: payload,
  })
}

export async function deleteChannelSlaRule(channelCode: string, ruleId: string): Promise<void> {
  await api(
    `/api/v1/channels/${encodeURIComponent(channelCode)}/sla-rules/${encodeURIComponent(ruleId)}`,
    { method: 'DELETE' },
  )
}

export async function listShipmentSlaAlerts(params: {
  scope?: 'todo' | 'all'
  riskLevel?: string
  alertType?: string
  status?: string
  hasException?: boolean
  exceptionCode?: string
  channelCode?: string
  carrierCode?: string
  customer?: string
  search?: string
  limit?: number
  offset?: number
}): Promise<ShipmentSlaAlertListResponse> {
  const query: Record<string, string> = {}
  if (params.scope) query.scope = params.scope
  if (params.riskLevel) query.riskLevel = params.riskLevel
  if (params.alertType) query.alertType = params.alertType
  if (params.status) query.status = params.status
  if (params.hasException != null) query.hasException = String(params.hasException)
  if (params.exceptionCode) query.exceptionCode = params.exceptionCode
  if (params.channelCode) query.channelCode = params.channelCode
  if (params.carrierCode) query.carrierCode = params.carrierCode
  if (params.customer) query.customer = params.customer
  if (params.search) query.search = params.search
  if (params.limit != null) query.limit = String(params.limit)
  if (params.offset != null) query.offset = String(params.offset)
  return api('/api/v1/shipment-sla-alerts', { query })
}

export async function getShipmentSlaSummary(): Promise<ShipmentSlaSummary> {
  return api('/api/v1/shipment-sla-alerts/summary')
}

export async function getShipmentSlaTodoNotifications(limit = 20): Promise<{
  items: ShipmentSlaAlert[]
  pendingCount: number
}> {
  return api('/api/v1/shipment-sla-alerts/todo-notifications', {
    query: { limit: String(limit) },
  })
}

export async function evaluateShipmentSlaAlerts(): Promise<{
  skipped: boolean
  reason?: string
  scanned: number
  created: number
  updated: number
  resolved: number
}> {
  return api('/api/v1/shipment-sla-alerts/evaluate', { method: 'POST' })
}

export async function followUpShipmentSlaAlert(alertId: string): Promise<void> {
  await api(`/api/v1/shipment-sla-alerts/${encodeURIComponent(alertId)}/follow-up`, {
    method: 'POST',
  })
}

/** @deprecated 使用 followUpShipmentSlaAlert */
export async function acknowledgeShipmentSlaAlert(alertId: string): Promise<void> {
  return followUpShipmentSlaAlert(alertId)
}

export async function resolveShipmentSlaAlert(alertId: string): Promise<void> {
  await api(`/api/v1/shipment-sla-alerts/${encodeURIComponent(alertId)}/resolve`, {
    method: 'POST',
  })
}

export async function ignoreShipmentSlaAlert(alertId: string): Promise<void> {
  await api(`/api/v1/shipment-sla-alerts/${encodeURIComponent(alertId)}/ignore`, {
    method: 'POST',
  })
}

export async function convertShipmentSlaAlert(
  alertId: string,
  payload: { exceptionCode?: string; note?: string },
): Promise<{ ok: boolean; eventId: string }> {
  return api(`/api/v1/shipment-sla-alerts/${encodeURIComponent(alertId)}/convert`, {
    method: 'POST',
    body: payload,
  })
}

export async function updateShipmentSlaAlertNote(alertId: string, note: string): Promise<void> {
  await api(`/api/v1/shipment-sla-alerts/${encodeURIComponent(alertId)}/note`, {
    method: 'PATCH',
    body: { note },
  })
}

function slaAlertTitle(riskLevel: SlaRiskLevel): string {
  if (riskLevel === 'severe_overdue') return '运输时效严重超时'
  if (riskLevel === 'overdue') return '运输时效已超时'
  return '运输时效即将超时'
}

function slaAlertMessage(alert: ShipmentSlaAlert): string {
  const days = alert.daysUntilDue
  if (days != null && days >= 0) {
    return `运单 ${alert.shipmentNo} · 预估截止 ${alert.dueDate}，剩余 ${days} 天。`
  }
  const overdue = alert.overdueDays || 0
  return `运单 ${alert.shipmentNo} · 预估截止 ${alert.dueDate}，已超期 ${overdue} 天。`
}

export function slaAlertAsCard(alert: ShipmentSlaAlert) {
  return {
    title: slaAlertTitle(alert.riskLevel),
    message: slaAlertMessage(alert),
    subtitle: alert.shipmentNo,
    customerName: alert.customer || '',
    ruleType: 'SLA_TRANSIT',
    severity: alert.severity,
    triggeredAt: alert.updatedTime || alert.createdTime,
  }
}
